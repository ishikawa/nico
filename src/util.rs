use std::cell::RefCell;
use std::rc::Rc;

pub fn wrap<T>(ty: T) -> Rc<RefCell<T>> {
    Rc::new(RefCell::new(ty))
}

pub trait UniqueNaming {
    fn next(&mut self) -> String;
}

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug)]
pub struct SequenceNaming {
    type_var_index: usize,
    prefix: String,
}

impl SequenceNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }
}

impl UniqueNaming for SequenceNaming {
    fn next(&mut self) -> String {
        let i = self.type_var_index;
        self.type_var_index += 1;
        format!("{}{}", self.prefix, i)
    }
}
