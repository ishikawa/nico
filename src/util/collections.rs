use crate::arena::BumpaloVec;
use std::cell::Ref;

#[derive(Debug)]
pub struct RefVecIter<'r, 'arena, T> {
    vec: Ref<'r, BumpaloVec<'arena, T>>,
}

impl<'r, 'arena, T> RefVecIter<'r, 'arena, T> {
    pub fn new(vec: Ref<'r, BumpaloVec<'arena, T>>) -> Self {
        Self { vec }
    }
}

impl<'a: 'r, 'r, 'arena, T> IntoIterator for &'a RefVecIter<'r, 'arena, T> {
    type IntoIter = std::slice::Iter<'r, T>;
    type Item = &'r T;

    fn into_iter(self) -> Self::IntoIter {
        self.vec.iter()
    }
}
