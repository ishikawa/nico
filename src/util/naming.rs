use std::collections::HashMap;

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug, Default)]
pub struct SequenceNaming {
    names: HashMap<String, i32>,
}

impl SequenceNaming {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next<S: AsRef<str>>(&mut self, name: S) -> String {
        let name = name.as_ref();
        let n = self.names.entry(name.to_string()).or_insert(0);
        let next = format!("{}.{}", name, n);

        *n += 1;
        next
    }
}

/// This struct produces a unique name while giving a sequential
/// number to the name passed to the next() method.
#[derive(Debug)]
pub struct PrefixNaming {
    type_var_index: usize,
    prefix: String,
}

impl PrefixNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }

    pub fn next(&mut self) -> String {
        let i = self.type_var_index;
        let next = format!("{}{}", self.prefix, i);

        self.type_var_index += 1;
        next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prefix_naming() {
        let mut naming = PrefixNaming::new("my_");

        assert_eq!(naming.next(), "my_0".to_string());
        assert_eq!(naming.next(), "my_1".to_string());
    }

    #[test]
    fn sequence_naming() {
        let mut naming = SequenceNaming::new();

        assert_eq!(naming.next("x"), "x.0".to_string());
        assert_eq!(naming.next("x"), "x.1".to_string());
        assert_eq!(naming.next("y"), "y.0".to_string());
    }
}
