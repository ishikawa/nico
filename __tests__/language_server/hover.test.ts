import { LanguageServer, spawn_server, LanguageServerAgent, Position } from "../util/lsp";
import { filterTestCases, TestCaseBase, readTestFileSync, getTestName } from "../util/testcase";

interface TestCase extends TestCaseBase {
  requests: Array<{ position: Position }>;
}

let server: LanguageServer | undefined;

beforeAll(async () => {
  server = await spawn_server({ rename: true });
});

afterAll(() => {
  if (server) {
    return server.stop();
  }
});

let cases: TestCase[] = [
  // struct field
  {
    // prettier-ignore
    input: [
      "struct A { a: i32 }",
      "let a = A { a: 100 }",
      "a.a"
    ].join("\n"),
    requests: [
      // struct A { a: i32 }
      //                ^
      {
        position: {
          line: 0,
          character: 15
        }
      },
      // let a = A { a: 100 }
      //             ^
      {
        position: {
          line: 1,
          character: 12
        }
      },
      // struct A { a: i32 }
      //            ^
      {
        position: {
          line: 0,
          character: 11
        }
      },
      // a.a
      //   ^
      {
        position: {
          line: 2,
          character: 2
        }
      },
      // let a = A { a: 100 }
      //     ^
      {
        position: {
          line: 1,
          character: 5
        }
      },
      // struct A { a: i32 }
      //         ^
      {
        position: {
          line: 0,
          character: 8
        }
      },
      // let a = A { a: 100 }
      //          ^
      {
        position: {
          line: 1,
          character: 9
        }
      }
    ]
  },
  // Type inference: binary expression
  {
    // prettier-ignore
    input: [
      "fun foo(n)",
      "  n + 1",
      "end",
      "foo(1)"
    ].join("\n"),
    requests: [
      // fun foo(n)
      //          ^
      {
        position: {
          line: 0,
          character: 9
        }
      },
      //   n + 1
      //    ^
      {
        position: {
          line: 1,
          character: 3
        }
      },
      // fun foo(n)
      //     ^
      {
        position: {
          line: 0,
          character: 5
        }
      },
      // foo(1)
      // ^
      {
        position: {
          line: 3,
          character: 0
        }
      }
    ]
  }
];

filterTestCases(cases).forEach((testCase, i) => {
  let src = readTestFileSync(testCase);
  let name = getTestName(testCase);

  testCase.requests.forEach(({ position }, j) => {
    test(`${i}: hover - \`${name}\` at ${position}`, async done => {
      const agent = new LanguageServerAgent(server!, { sequence: [i, j] });

      // Open document and no compilation errors
      const diagnostics = await agent.openDocument(name, src);
      expect(diagnostics).toHaveLength(0);

      const response = await agent.hover(name, position);
      expect(response).toMatchSnapshot();

      done();
    });
  });
});
