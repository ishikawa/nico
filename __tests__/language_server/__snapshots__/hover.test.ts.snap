// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1000,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
i32
\`\`\`
---
The 32-bit signed integer type.",
    },
    "range": Object {
      "end": Object {
        "character": 17,
        "line": 0,
      },
      "start": Object {
        "character": 14,
        "line": 0,
      },
    },
  },
}
`;

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 2`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
A.a: i32
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 13,
        "line": 1,
      },
      "start": Object {
        "character": 12,
        "line": 1,
      },
    },
  },
}
`;

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 3`] = `
Object {
  "error": null,
  "id": 1002,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
A.a: i32
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 12,
        "line": 0,
      },
      "start": Object {
        "character": 11,
        "line": 0,
      },
    },
  },
}
`;

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 4`] = `
Object {
  "error": null,
  "id": 1003,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
A.a: i32
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 2,
      },
      "start": Object {
        "character": 2,
        "line": 2,
      },
    },
  },
}
`;

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 5`] = `
Object {
  "error": null,
  "id": 1004,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
let a: A
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 5,
        "line": 1,
      },
      "start": Object {
        "character": 4,
        "line": 1,
      },
    },
  },
}
`;

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 6`] = `
Object {
  "error": null,
  "id": 1005,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
struct A
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 8,
        "line": 0,
      },
      "start": Object {
        "character": 7,
        "line": 0,
      },
    },
  },
}
`;

exports[`0: hover - \`struct A { a: i32 }
let a = A { a: 100 }
a.a\` at [object Object] 7`] = `
Object {
  "error": null,
  "id": 1006,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
struct A
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 9,
        "line": 1,
      },
      "start": Object {
        "character": 8,
        "line": 1,
      },
    },
  },
}
`;

exports[`1: hover - \`fun foo(n)
  n + 1
end\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1010,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
(parameter) n: i32
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 9,
        "line": 0,
      },
      "start": Object {
        "character": 8,
        "line": 0,
      },
    },
  },
}
`;

exports[`1: hover - \`fun foo(n)
  n + 1
end\` at [object Object] 2`] = `
Object {
  "error": null,
  "id": 1011,
  "jsonrpc": "2.0",
  "result": Object {
    "contents": Object {
      "kind": "markdown",
      "value": "\`\`\`nico
(parameter) n: i32
\`\`\`",
    },
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
  },
}
`;
